src/Microsoft.Azure.WebJobs.Host/Blobs/Listeners/ScanBlobScanLogHybridPollingStrategy.cs

```diff
@@ -0,0 +1,162 @@
+using Microsoft.Azure.WebJobs.Host.Storage.Blob;
+using Microsoft.Azure.WebJobs.Host.Timers;
+using Microsoft.WindowsAzure.Storage;
+
+namespace Microsoft.Azure.WebJobs.Host.Blobs.Listeners
+{
+    internal sealed class ScanBlobScanLogHybridPollingStrategy : IBlobListenerStrategy
+    {
+        private static readonly TimeSpan TwoSeconds = TimeSpan.FromSeconds(2);
+        private enum PollStrategy { ScanBlobs, ScanLogs, NotDetermined };
+        private int _scanBlobThreshold = 5000;
+        private readonly IDictionary<IStorageBlobContainer, PollStrategy> _pollStrategies;
+        private readonly IDictionary<IStorageBlobContainer, ICollection<ITriggerExecutor<IStorageBlob>>> _registrations;
+        private ScanContainersStrategy _scanContainerStrategy;
+        private PollLogsStrategy _pollLogStrategy;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private bool _disposed;
+
+        public ScanBlobScanLogHybridPollingStrategy() : base()
+        {
+            _registrations = new Dictionary<IStorageBlobContainer, ICollection<ITriggerExecutor<IStorageBlob>>>(
+                new StorageBlobContainerComparer());
+            _pollStrategies = new Dictionary<IStorageBlobContainer, PollStrategy>(
+                new StorageBlobContainerComparer());
+            _scanContainerStrategy = new ScanContainersStrategy();
+            _pollLogStrategy = new PollLogsStrategy();
+            _cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public void Start()
+        {
+            ThrowIfDisposed();
+            // can not start the 2 strategies till all regiesterations are done
+            ScnaContainersAndRegisterWithStrategy(_cancellationTokenSource.Token);
+            _scanContainerStrategy.Start();
+            _pollLogStrategy.Start();
+        }
+
+        public void Cancel()
+        {
+            ThrowIfDisposed();
+            _scanContainerStrategy.Cancel();
+            _pollLogStrategy.Cancel();
+            _cancellationTokenSource.Cancel();
+        }
+
+        public Task RegisterAsync(IStorageBlobContainer container, ITriggerExecutor<IStorageBlob> triggerExecutor, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            // Register and Execute are not concurrency-safe.
+            // Avoiding calling Register while Execute is running is the caller's responsibility.
+
+            ICollection<ITriggerExecutor<IStorageBlob>> containerRegistrations;
+            if (_registrations.ContainsKey(container))
+            {
+                containerRegistrations = _registrations[container];
+            }
+            else
+            {
+                containerRegistrations = new List<ITriggerExecutor<IStorageBlob>>();
+                _registrations.Add(container, containerRegistrations);
+            }
+
+            containerRegistrations.Add(triggerExecutor);
+            return Task.FromResult(0);
```

Comment:
it's just a convention for returning a completed task since this is not an 'async' method. Ideally, BCL would have had "Task.Completed" 

---
In reply to: [54306785](https://github.com/Azure/azure-webjobs-sdk/pull/637#discussion_r54306785) [](ancestors = 54306785)